### PR TITLE
(SERVER) Auth0 Token usage and Role Base Access Control

### PR DESCRIPTION
**Pull request by bseptember**
**(SERVER) Auth0 Token usage and Role Base Access Control**

**Overview**:

Authorization using the access token from Auth0 which includes roles & permissions found the jwt. Instead of using the generate token function from the code. This will prevent any user without an access token from Auth0 AND a role assigned which can't be tampered with otherwise the public key will not be able to verify it since it is signed by Auth0's private key.

Added  auth service that uses a management token from Auth0 api to make request to Auth0's user database. These include add, delete, update a role, as well as assign, remove a role from a user among others.

Eg. When a user with privileges changes a role of another user, the auth service sends an api request to the Auth0 management api to change the role. The database is still updated as per normal.

Table EmployeeRole added authRoleId & Table Employee added authUserId.

**Note**:

This branch needs the Client RGO branch with the same name to work.

Future:

Some endpoint still need the [Authorize] or policy added to it.

**Known issues**:

Permissions from auth0 are not yet linked to the db. 
Unit tests need to be updated.

### CODE REVIEW

File: HRIS.Services/Interfaces/IAuthService.cs
Comment (MatthewSchoeman) [inline HRIS.Services/Interfaces/IAuthService.cs:1]: Please can documentation be added as it is present in the other interface document

File: RR.App.Tests/Controllers/HRIS/PropertyAccessControllerUnitTests.cs
Comment (MatthewSchoeman) [inline RR.App.Tests/Controllers/HRIS/PropertyAccessControllerUnitTests.cs:1]: Please remove the commented code

**Review by EduanO — APPROVED** [review]

File: HRIS.Services.Tests/Services/AuthServiceUnitTest.cs [hunk 1]
```diff
@@ -57,232 +55,8 @@ public AuthServiceUnitTest()
...
-        var email = "dm@retrorabbit.co.za";
-        var expectedRoles = new Dictionary<string, List<string>>
-            { { "Admin", new List<string> { "Create", "Read", "Update", "Delete" } } };
-        _roleAccessLinkService.Setup(x => x.GetRoleByEmployee(email)).ReturnsAsync(expectedRoles);
-
-        var roles = await _authService.GetUserRoles(email);
-
-        Assert.Equal(expectedRoles, roles);
-    }
+    // TODO: update this unit test
```
Comment (tbddorothy) [inline HRIS.Services.Tests/Services/AuthServiceUnitTest.cs:61]: Why did you delete this unit test?

File: HRIS.Services/Interfaces/IAuthService.cs [hunk 2]
```diff
@@ -1,41 +1,41 @@
...
+
+    Task<IList<User>> GetUsersByEmailAsync(string email);
+
+    Task<IPagedList<AssignedUser>> GetUsersByRoleAsync(string roleId);
+
+    Task<IPagedList<Role>> GetUserRolesAsync(string userId);
+
+    Task<bool> AddRoleToUserAsync(string userId, string roleId);
+
+    Task<bool> RemoveRoleFromUserAsync(string userId, string roleId);
```
Comment (tbddorothy) [inline HRIS.Services/Interfaces/IAuthService.cs:19]: Docs Missing

File: HRIS.Services.Tests/Services/AuthServiceUnitTest.cs (hunk 1 above)
Comment (bseptember) [reply HRIS.Services.Tests/Services/AuthServiceUnitTest.cs:61]: The Auth Service has been completely changed. There is no proper unit test written for the revamped service yet.

File: HRIS.Services/Interfaces/IAuthService.cs (hunk 2 above)
Comment (bseptember) [reply HRIS.Services/Interfaces/IAuthService.cs:19]: Will add

**Review by Wimmas — DISMISSED** [review]

File: RR.App.Tests/Controllers/HRIS/PropertyAccessControllerUnitTests.cs
Comment (bseptember) [reply RR.App.Tests/Controllers/HRIS/PropertyAccessControllerUnitTests.cs:1]: TODOs left for future tickets

**Review by EduanO — DISMISSED** [review]

**Review by MatthewSchoeman — APPROVED** [review]

**Review by tbddorothy — APPROVED** [review]

**Review by Juanro-Minne — DISMISSED** [review]

**Review by tbddorothy — APPROVED** [review]

**Review by ReinoRabbit — DISMISSED** [review]

**Review by MatthewSchoeman — DISMISSED** [review]

**Review by Juanro-Minne — APPROVED** [review]

**Review by EduanO — APPROVED** [review]

**Review by ReinoRabbit — APPROVED** [review]
